docs/Toy/toy/rewrites/lower_memref_riscv.py

```diff
@@ -0,0 +1,125 @@
+from math import prod
+from typing import Any, Sequence, cast
+
+from xdsl.dialects import memref, riscv
+from xdsl.dialects.builtin import ModuleOp, UnrealizedConversionCastOp
+from xdsl.ir.core import MLContext, SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+from .helpers import cast_value_to_register
+
+
+class LowerMemrefAllocOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.Alloc, rewriter: PatternRewriter):
+        assert isinstance(op.memref.type, memref.MemRefType)
+        memref_typ = cast(memref.MemRefType[Any], op.memref.type)
+        size = prod(op.memref.type.get_shape())
+        rewriter.replace_matched_op(
+            [
+                size_op := riscv.LiOp(size, comment="memref alloc size"),
+                alloc_op := riscv.CustomAssemblyInstructionOp(
+                    "buffer.alloc",
+                    (size_op.rd,),
+                    (riscv.IntRegisterType.unallocated(),),
+                ),
+                UnrealizedConversionCastOp.get(alloc_op.results, (memref_typ,)),
+            ]
+        )
+
+
+def insert_shape_ops(
+    mem: SSAValue,
+    indices: Sequence[SSAValue],
+    shape: Sequence[int],
+    rewriter: PatternRewriter,
+) -> SSAValue:
+    """
+    Returns ssa value representing pointer into the memref at given indices.
+    """
+    assert len(shape) == len(indices)
+
+    if len(shape) == 1:
+        rewriter.insert_op_before_matched_op(
+            [
+                ptr := riscv.AddOp(mem, indices[0]),
+            ]
+        )
+    elif len(shape) == 2:
+        rewriter.insert_op_before_matched_op(
+            [
+                cols := riscv.LiOp(shape[1]),
+                row_offset := riscv.MulOp(cols, indices[0]),
+                offset := riscv.AddOp(row_offset, indices[1]),
+                offset_bytes := riscv.SlliOp(offset, 2, comment="mutiply by elm size"),
+                ptr := riscv.AddOp(mem, offset_bytes),
+            ]
+        )
+    else:
+        raise NotImplementedError(f"Unsupported memref shape {shape}")
```

Comment:
Feel like it can be rewriting using pattern matching which will be more concise and readable.

Reply:
Something like this works, tested on my machine but you may rename the variable name for readability.
```suggestion
    match indices:
        case [el]:
            rewriter.insert_op_before_matched_op(
                [
                    ptr := riscv.AddOp(mem, el),
                ]
            )
        case [first, second]:
            rewriter.insert_op_before_matched_op(
                [
                    cols := riscv.LiOp(shape[1]),
                    row_offset := riscv.MulOp(cols, first),
                    offset := riscv.AddOp(row_offset, second),
                    offset_bytes := riscv.SlliOp(
                        offset, 2, comment="mutiply by elm size"
                    ),
                    ptr := riscv.AddOp(mem, offset_bytes),
                ]
            )
        case _:
            raise NotImplementedError(f"Unsupported memref shape {shape}")
```